File: app/src/main/java/com/example/policyfolio/UI/Fragments/NavigationActionFragments/AddPolicyDetailsFragment.java
package com.example.policyfolio.UI.Fragments.NavigationActionFragments;


import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.provider.MediaStore;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.example.policyfolio.Util.Constants;
import com.example.policyfolio.Repo.Database.DataClasses.Nominee;
import com.example.policyfolio.R;
import com.example.policyfolio.Util.Adapters.BasicDropdownNomineeAdapter;
import com.example.policyfolio.Util.Adapters.BasicDropdownTextAdapter;
import com.example.policyfolio.Util.CallBackListeners.NavigationCallbacks.AddPolicyCallback;
import com.example.policyfolio.ViewModels.NavigationViewModels.AddViewModel;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 */
public class AddPolicyDetailsFragment extends Fragment implements BasicDropdownTextAdapter.ParentCallback, BasicDropdownNomineeAdapter.ParentCallback {

    private View rootView;
    private AddViewModel viewModel;

    private TextView insuranceProvider;
    private TextView policyNumber;
    private EditText coverAmount;

    private FrameLayout premiumFrame;
    private TextView premiumText;
    private RecyclerView premiumChoice;
    private BasicDropdownTextAdapter premiumAdapter;
    private String[] premiums;

    private EditText premiumAmount;

    private LinearLayout premiumDatePickerLayout;
    private TextView premiumDate;
    private ImageView premiumDateImage;
    private Long premiumDateEpoch;

    private LinearLayout matureDatePickerLayout;
    private TextView matureDate;
    private ImageView matureDateImage;
    private Long matureDateEpoch;


    private FrameLayout nomineeFrame;
    private TextView nomineeText;
    private RecyclerView nomineeChoice;
    private ArrayList<Nominee> nominees;
    private BasicDropdownNomineeAdapter nomineeAdapter;

    private LinearLayout addPolicy;
    private TextView documentAddText;
    private ImageView documentAddImage;
    private TextView optional1;

    private LinearLayout clickPolicy;
    private TextView optional2;

    private Button done;

    private AddPolicyCallback callback;

    public AddPolicyDetailsFragment() {
        // Required empty public constructor
    }

    public AddPolicyDetailsFragment(AddPolicyCallback callback){
        this.callback = callback;
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_add_policy_details, container, false);
        viewModel = ViewModelProviders.of(getActivity()).get(AddViewModel.class);
        viewModel.initiateRepo(getContext());

        insuranceProvider = rootView.findViewById(R.id.insurance_provider);
        policyNumber = rootView.findViewById(R.id.policy_number);
        coverAmount = rootView.findViewById(R.id.cover_amount);

        premiumFrame = rootView.findViewById(R.id.frequency_frame);
        premiumText = rootView.findViewById(R.id.frequency_text);
        premiumChoice = rootView.findViewById(R.id.frequency);
        premiums = getResources().getStringArray(R.array.premium_frequency);

        premiumAmount = rootView.findViewById(R.id.premium_amount);

        premiumDatePickerLayout = rootView.findViewById(R.id.date_picker);
        premiumDate = rootView.findViewById(R.id.date);
        premiumDateImage = rootView.findViewById(R.id.date_image);

        matureDatePickerLayout = rootView.findViewById(R.id.date_picker2);
        matureDate = rootView.findViewById(R.id.date2);
        matureDateImage = rootView.findViewById(R.id.date_image2);

        nomineeFrame = rootView.findViewById(R.id.nominee_frame);
        nomineeText = rootView.findViewById(R.id.nominee_text);
        nomineeChoice = rootView.findViewById(R.id.nominee);
        nominees = new ArrayList<>();

        addPolicy = rootView.findViewById(R.id.add_document);
        documentAddText = rootView.findViewById(R.id.add_document_text);
        documentAddImage = rootView.findViewById(R.id.document_add_image);
        optional1 = rootView.findViewById(R.id.optional1);

        clickPolicy = rootView.findViewById(R.id.click_document);
        optional2 = rootView.findViewById(R.id.optional2);

        done = rootView.findViewById(R.id.done);

        setAdapters();
        setDefaults();

        return rootView;
    }

    private void setAdapters() {
        coverAmount.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String text = s.toString();
                if(text.length()>0){
                    coverAmount.setPadding(0,4,0,4);
                    coverAmount.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                }
                else {
                    coverAmount.setPadding(24,4,32,4);
                    coverAmount.setBackground(getResources().getDrawable(R.drawable.text_background_8dp));
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        premiumAdapter = new BasicDropdownTextAdapter(getContext(),premiums, this,Constants.DropDownType.PREMIUM_FREQUENCY);
        premiumChoice.setAdapter(premiumAdapter);
        premiumChoice.setLayoutManager(new LinearLayoutManager(getActivity(),RecyclerView.VERTICAL,false));

        premiumFrame.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                coverAmount.clearFocus();
                premiumAmount.clearFocus();

                premiumFrame.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                premiumFrame.setPadding(0,4,0,4);
                premiumText.setVisibility(View.GONE);
                premiumChoice.setVisibility(View.VISIBLE);
            }
        });

        premiumAmount.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String text = s.toString();
                if(text.length()>0){
                    premiumAmount.setPadding(0,4,0,4);
                    premiumAmount.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                }
                else {
                    premiumAmount.setPadding(24,4,32,4);
                    premiumAmount.setBackground(getResources().getDrawable(R.drawable.text_background_8dp));
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        premiumDatePickerLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                coverAmount.clearFocus();
                premiumAmount.clearFocus();

                final Dialog dialog = new Dialog(getContext(),R.style.DialogTheme);
                dialog.setContentView(R.layout.date_picker);
                dialog.setTitle("");
                DatePicker datePicker = dialog.findViewById(R.id.date_picker);
                final Calendar calendar=Calendar.getInstance();
                calendar.setTimeInMillis(premiumDateEpoch*1000);
                datePicker.init(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH), new DatePicker.OnDateChangedListener() {
                    @Override
                    public void onDateChanged(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                        calendar.set(year,monthOfYear,dayOfMonth);
                        premiumDateEpoch = (calendar.getTimeInMillis())/1000;
                        viewModel.setPremiumDateEpoch(premiumDateEpoch);
                        premiumDate.setText(Constants.Time.DATE_FORMAT.format(premiumDateEpoch*1000));
                        premiumDate.setPadding(0,4,0,4);
                        premiumDate.setTextColor(getResources().getColor(R.color.colorPrimaryDark));
                        premiumDateImage.setVisibility(View.GONE);
                        premiumDatePickerLayout.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                    }
                });
                dialog.show();
            }
        });

        matureDatePickerLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                coverAmount.clearFocus();
                premiumAmount.clearFocus();


                final Dialog dialog = new Dialog(getContext(),R.style.DialogTheme);
                dialog.setContentView(R.layout.date_picker);
                dialog.setTitle("");
                DatePicker datePicker = dialog.findViewById(R.id.date_picker);
                final Calendar calendar=Calendar.getInstance();
                calendar.setTimeInMillis(matureDateEpoch*1000);
                datePicker.init(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH), new DatePicker.OnDateChangedListener() {
                    @Override
                    public void onDateChanged(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                        calendar.set(year,monthOfYear,dayOfMonth);
                        matureDateEpoch = (calendar.getTimeInMillis())/1000;
                        viewModel.setMatureDateEpoch(matureDateEpoch);
                        matureDate.setText(Constants.Time.DATE_FORMAT.format(matureDateEpoch*1000));
                        matureDate.setPadding(0,4,0,4);
                        matureDate.setTextColor(getResources().getColor(R.color.colorPrimaryDark));
                        matureDateImage.setVisibility(View.GONE);
                        matureDatePickerLayout.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                    }
                });
                dialog.show();
            }
        });

        nomineeAdapter = new BasicDropdownNomineeAdapter(getContext(),nominees,this);
        nomineeChoice.setAdapter(nomineeAdapter);
        nomineeChoice.setLayoutManager(new LinearLayoutManager(getActivity(),RecyclerView.VERTICAL,false));

        nomineeFrame.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                coverAmount.clearFocus();
                premiumAmount.clearFocus();

                if(nominees.size()!=0) {
                    nomineeFrame.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                    nomineeText.setVisibility(View.GONE);
                    nomineeChoice.setVisibility(View.VISIBLE);
                }
                else
                    Toast.makeText(getContext(),"No Nominees exist for the User",Toast.LENGTH_LONG).show();
            }
        });

        addPolicy.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                coverAmount.clearFocus();
                premiumAmount.clearFocus();

                callback.addPolicyImage();
            }
        });

        clickPolicy.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                coverAmount.clearFocus();
                premiumAmount.clearFocus();

                callback.clickPolicyImage();
            }
        });

        done.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                coverAmount.clearFocus();
                premiumAmount.clearFocus();

                if(!(premiumAmount.getText().toString().equals("") || premiumDateEpoch == null || premiumText.getVisibility()==View.GONE || premiumText.getCurrentTextColor()!=getResources().getColor(R.color.colorPrimaryDark))) {
                    viewModel.setPremiumAmount(premiumAmount.getText().toString());
                    viewModel.setCoverAmount(coverAmount.getText().toString());
                    callback.done();
                }
                else
                    Toast.makeText(getContext(),"Information Incomplete",Toast.LENGTH_LONG).show();
            }
        });
    }

    private void setDefaults() {
        insuranceProvider.setText(viewModel.getProvider().getName());
        policyNumber.setText(viewModel.getPolicyNumber());
        viewModel.fetchNominees().observe(this, new Observer<List<Nominee>>() {
            @Override
            public void onChanged(List<Nominee> nominees) {
                AddPolicyDetailsFragment.this.nominees.clear();
                AddPolicyDetailsFragment.this.nominees.addAll(nominees);
                nomineeAdapter.notifyDataSetChanged();
            }
        });
        premiumDateEpoch = System.currentTimeMillis()/1000;
        matureDateEpoch = System.currentTimeMillis()/1000;
    }

    @Override
    public void setValue(int position, int type) {
        coverAmount.clearFocus();
        premiumAmount.clearFocus();
        switch (type){
            case Constants.DropDownType.PREMIUM_FREQUENCY:
                premiumText.setText(premiums[position]);
                premiumText.setTextColor(getResources().getColor(R.color.colorPrimaryDark));
                premiumText.setPadding(0,4,0,4);
                premiumChoice.setVisibility(View.GONE);
                premiumText.setVisibility(View.VISIBLE);
                viewModel.setPremiumFrequency(position);
                break;
            case Constants.DropDownType.NOMINEE:
                String[] relations = getResources().getStringArray(R.array.relationship_array);
                nomineeText.setText(nominees.get(position).getName() + ", " + relations[nominees.get(position).getRelation()]);
                nomineeText.setTextColor(getResources().getColor(R.color.colorPrimaryDark));
                nomineeText.setPadding(0,4,0,4);
                nomineeChoice.setVisibility(View.GONE);
                nomineeText.setVisibility(View.VISIBLE);
                viewModel.setNominee(nominees.get(position));
                break;
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode == Constants.PermissionAndRequests.PICK_IMAGE_REQUEST && resultCode == Activity.RESULT_OK && data!=null) {
            Uri selectedImage = data.getData();
            String[] filePathColumn = { MediaStore.Images.Media.DATA };

            Cursor cursor = getActivity().getContentResolver().query(selectedImage, filePathColumn, null, null, null);
            cursor.moveToFirst();

            int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
            String picturePath = cursor.getString(columnIndex);
            cursor.close();

            Bitmap bmp = BitmapFactory.decodeFile(picturePath);
            Log.e("BITMAP",bmp+"");
            documentAddImage.setVisibility(View.GONE);
            documentAddText.setVisibility(View.GONE);
            clickPolicy.setVisibility(View.GONE);
            optional2.setVisibility(View.GONE);
            optional1.setText("Policy Document");
            viewModel.saveImage(bmp).observe(this, new Observer<String>() {
                @Override
                public void onChanged(String s) {
                   if (s != null)
                       viewModel.setPhotoUrl(s);
                   else
                       Toast.makeText(getContext(), "Error uploading Image", Toast.LENGTH_LONG).show();
                }
            });
        }
        if(requestCode == Constants.PermissionAndRequests.CAPTURE_IMAGE_REQUEST && resultCode == Activity.RESULT_OK && data!=null){
            Bitmap bmp = (Bitmap) data.getExtras().get("data");
            documentAddImage.setVisibility(View.GONE);
            documentAddText.setVisibility(View.GONE);
            clickPolicy.setVisibility(View.GONE);
            optional2.setVisibility(View.GONE);
            optional1.setText("Policy Document");
                viewModel.saveImage(bmp).observe(this, new Observer<String>() {
                    @Override
                    public void onChanged(String s) {
                        if (s != null)
                            viewModel.setPhotoUrl(s);
                        else
                            Toast.makeText(getContext(), "Error uploading Image", Toast.LENGTH_LONG).show();
                    }
                });
        }
    }
}

File: app/src/main/java/com/example/policyfolio/ui/addpolicy/AddPolicyDetailsFragment.kt
package com.example.policyfolio.ui.addpolicy


import android.app.Activity
import android.app.DatePickerDialog
import android.app.Dialog
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle

import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders

import android.provider.MediaStore
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.DatePicker
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.view.isGone
import androidx.core.view.isVisible

import com.example.policyfolio.util.Constants
import com.example.policyfolio.data.local.classes.Nominee
import com.example.policyfolio.R
import com.example.policyfolio.ui.adapters.ListAdapters.BasicDropdownNomineeAdapter
import com.example.policyfolio.ui.adapters.ListAdapters.BasicDropdownTextAdapter
import com.example.policyfolio.viewmodels.AddPolicyViewModel
import com.wajahatkarim3.easyvalidation.core.view_ktx.nonEmpty
import com.wajahatkarim3.easyvalidation.core.view_ktx.validator

import java.io.ByteArrayOutputStream
import java.util.ArrayList
import java.util.Calendar

/**
 * A simple [Fragment] subclass.
 */
class AddPolicyDetailsFragment(private val callback: AddPolicyCallback) : Fragment(), BasicDropdownTextAdapter.ParentCallback, BasicDropdownNomineeAdapter.ParentCallback {

    private var rootView: View? = null
    private var viewModel: AddPolicyViewModel? = null

    private var insuranceProvider: TextView? = null
    private var policyNumber: TextView? = null
    private var coverAmount: EditText? = null
    private var coverError: TextView? = null

    private var premiumFrame: FrameLayout? = null
    private var premiumText: TextView? = null
    private var premiumAdapter: BasicDropdownTextAdapter? = null
    private var premiums: Array<String>? = null
    private var frequencyError: TextView? = null

    private var premiumAmount: EditText? = null
    private var premiumError: TextView? = null

    private var dueDatePickerLayout: LinearLayout? = null
    private var dueDate: TextView? = null
    private var dueDateImage: ImageView? = null
    private var premiumDateEpoch: Long? = null
    private var dueError: TextView? = null

    private var matureDatePickerLayout: LinearLayout? = null
    private var matureDate: TextView? = null
    private var matureDateImage: ImageView? = null
    private var matureDateEpoch: Long? = null
    private var matureError: TextView? = null

    private var nomineeFrame: FrameLayout? = null
    private var nomineeText: TextView? = null
    private var nominees: ArrayList<Nominee>? = null
    private var nomineeAdapter: BasicDropdownNomineeAdapter? = null
    private var nomineeError: TextView? = null

    private var addPolicy: LinearLayout? = null
    private var documentAddText: TextView? = null
    private var documentAddImage: ImageView? = null
    private var optional1: TextView? = null

    private var clickPolicy: LinearLayout? = null
    private var optional2: TextView? = null

    private var done: Button? = null


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.fragment_add_policy_details, container, false)
        viewModel = ViewModelProviders.of(activity!!).get(AddPolicyViewModel::class.java)
        viewModel!!.initiateRepo(context)

        insuranceProvider = rootView!!.findViewById(R.id.insurance_provider)
        policyNumber = rootView!!.findViewById(R.id.policy_number)
        coverAmount = rootView!!.findViewById(R.id.cover_amount)
        coverError = rootView!!.findViewById(R.id.cover_empty)

        premiumFrame = rootView!!.findViewById(R.id.frequency_frame)
        premiumText = rootView!!.findViewById(R.id.frequency_text)
        premiums = resources.getStringArray(R.array.premium_frequency)
        frequencyError = rootView!!.findViewById(R.id.frequency_empty)

        premiumAmount = rootView!!.findViewById(R.id.premium_amount)
        premiumError = rootView!!.findViewById(R.id.premium_empty)

        dueDatePickerLayout = rootView!!.findViewById(R.id.date_picker)
        dueDate = rootView!!.findViewById(R.id.date)
        dueDateImage = rootView!!.findViewById(R.id.date_image)
        dueError = rootView!!.findViewById(R.id.due_empty)

        matureDatePickerLayout = rootView!!.findViewById(R.id.date_picker2)
        matureDate = rootView!!.findViewById(R.id.date2)
        matureDateImage = rootView!!.findViewById(R.id.date_image2)
        matureError = rootView!!.findViewById(R.id.mature_empty)

        nomineeFrame = rootView!!.findViewById(R.id.nominee_frame)
        nomineeText = rootView!!.findViewById(R.id.nominee_text)
        nominees = ArrayList()
        nomineeError = rootView!!.findViewById(R.id.nominee_empty)

        addPolicy = rootView!!.findViewById(R.id.add_document)
        documentAddText = rootView!!.findViewById(R.id.add_document_text)
        documentAddImage = rootView!!.findViewById(R.id.document_add_image)
        optional1 = rootView!!.findViewById(R.id.optional1)

        clickPolicy = rootView!!.findViewById(R.id.click_document)
        optional2 = rootView!!.findViewById(R.id.optional2)

        done = rootView!!.findViewById(R.id.done)

        setAdapters()
        setDefaults()

        return rootView
    }

    private fun setAdapters() {
        coverAmount!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val text = s.toString()
                if (text.isNotEmpty()) {
                    coverAmount!!.setPadding(0, 4, 0, 4)
                    coverAmount!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
                } else {
                    coverAmount!!.setPadding(24, 4, 32, 4)
                    coverAmount!!.background = resources.getDrawable(R.drawable.text_background_grey_8dp)
                }
            }

            override fun afterTextChanged(s: Editable) {

            }
        })

        premiumAdapter = BasicDropdownTextAdapter(context, premiums, this, Constants.ListTypes.PREMIUM_FREQUENCY, resources.getColor(R.color.colorAccent))
        viewModel!!.premiumFrequency = -1
        premiumFrame!!.setOnClickListener {
            premiumFrame!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
            premiumFrame!!.setPadding(0, 4, 0, 4)
            callback.openListSheet(Constants.ListTypes.PREMIUM_FREQUENCY, premiumAdapter)
        }

        premiumAmount!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val text = s.toString()
                if (text.isNotEmpty()) {
                    premiumAmount!!.setPadding(0, 4, 0, 4)
                    premiumAmount!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
                } else {
                    premiumAmount!!.setPadding(24, 4, 32, 4)
                    premiumAmount!!.background = resources.getDrawable(R.drawable.text_background_grey_8dp)
                }
            }

            override fun afterTextChanged(s: Editable) {

            }
        })

        dueDatePickerLayout!!.setOnClickListener {
            val calendar = Calendar.getInstance()
            calendar.timeInMillis = premiumDateEpoch!! * 1000

            val dateSetListener = DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
                calendar.set(year, monthOfYear, dayOfMonth)
                premiumDateEpoch = calendar.timeInMillis / 1000
                viewModel!!.premiumDateEpoch = premiumDateEpoch
                dueDate!!.text = Constants.Time.DATE_FORMAT.format(premiumDateEpoch!! * 1000)
                dueDate!!.setPadding(0, 4, 0, 4)
                dueDate!!.setTextColor(resources.getColor(R.color.colorPrimaryDark))
                dueDateImage!!.visibility = View.GONE
                dueDatePickerLayout!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
            }

            DatePickerDialog(context,R.style.MyDatePickerDialogTheme, dateSetListener,
                    calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
                    calendar.get(Calendar.DAY_OF_MONTH)).show()
        }

        matureDatePickerLayout!!.setOnClickListener {
            val calendar = Calendar.getInstance()
            calendar.timeInMillis = premiumDateEpoch!! * 1000

            val dateSetListener = DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
                calendar.set(year, monthOfYear, dayOfMonth)
                matureDateEpoch = calendar.timeInMillis / 1000
                viewModel!!.matureDateEpoch = matureDateEpoch
                matureDate!!.text = Constants.Time.DATE_FORMAT.format(matureDateEpoch!! * 1000)
                matureDate!!.setPadding(0, 4, 0, 4)
                matureDate!!.setTextColor(resources.getColor(R.color.colorPrimaryDark))
                matureDateImage!!.visibility = View.GONE
                matureDatePickerLayout!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
            }

            DatePickerDialog(context,R.style.MyDatePickerDialogTheme, dateSetListener,
                    calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
                    calendar.get(Calendar.DAY_OF_MONTH)).show()
        }

        nomineeAdapter = BasicDropdownNomineeAdapter(context, nominees, this)

        nomineeFrame!!.setOnClickListener {
            if (nominees!!.size != 0) {
                nomineeFrame!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
                callback.openListSheet(Constants.ListTypes.NOMINEE, nomineeAdapter)
            } else
                callback.showSnackbar("No Nominees exist for the User")
        }

        addPolicy!!.setOnClickListener {
            callback.addPolicyImage()
        }

        clickPolicy!!.setOnClickListener {
            callback.clickPolicyImage()
        }

        done!!.setOnClickListener {
            var isComplete:Boolean = true
            coverAmount!!.validator()
                    .nonEmpty()
                    .addErrorCallback { message ->
                        coverError!!.text = message
                        coverError!!.setTextColor(resources!!.getColor(R.color.red))
                        isComplete = false
                    }
                    .addSuccessCallback {
                        coverError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
                    }.check()

            if(viewModel!!.premiumFrequency == -1){
                frequencyError!!.setTextColor(resources!!.getColor(R.color.red))
                isComplete = false
            } else{
                frequencyError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
            }

            premiumAmount!!.validator()
                    .nonEmpty()
                    .addErrorCallback { message ->
                        premiumError!!.text = message
                        premiumError!!.setTextColor(resources!!.getColor(R.color.red))
                        isComplete = false
                    }
                    .addSuccessCallback {
                        premiumError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
                    }.check()

            if(dueDateImage!!.isGone){
                dueError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
            } else{
                dueError!!.setTextColor(resources!!.getColor(R.color.red))
                isComplete = false
            }

            if(matureDateImage!!.isGone){
                matureError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
            } else{
                matureError!!.setTextColor(resources!!.getColor(R.color.red))
                isComplete = false
            }

            if(viewModel!!.nominee == null){
                nomineeError!!.setTextColor(resources!!.getColor(R.color.red))
                isComplete = false
            } else{
                nomineeError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
            }

            if(isComplete){
                viewModel!!.coverAmount = coverAmount!!.text.toString()
                viewModel!!.premiumAmount = premiumAmount!!.text.toString()
                callback.done()
            } else {
                callback.showSnackbar("Incomplete Information!")
            }
        }
    }

    private fun setDefaults() {
        insuranceProvider!!.text = viewModel!!.provider.name
        policyNumber!!.text = viewModel!!.policyNumber
        viewModel!!.fetchNominees().observe(this, Observer { results ->
            nominees!!.clear()
            nominees!!.addAll(results)
            nomineeAdapter!!.notifyDataSetChanged()
        })
        premiumDateEpoch = System.currentTimeMillis() / 1000
        matureDateEpoch = System.currentTimeMillis() / 1000
    }

    override fun setValue(position: Int, type: Int) {
        coverAmount!!.clearFocus()
        premiumAmount!!.clearFocus()
        callback.closeListSheet()
        when (type) {
            Constants.ListTypes.PREMIUM_FREQUENCY -> {
                premiumText!!.text = premiums!![position]
                premiumText!!.setTextColor(resources.getColor(R.color.colorPrimaryDark))
                premiumText!!.setPadding(0, 4, 0, 4)
                viewModel!!.premiumFrequency = position
            }
            Constants.ListTypes.NOMINEE -> {
                val relations = resources.getStringArray(R.array.relationship_array)
                nomineeText!!.text = nominees!![position].name + ", " + relations[nominees!![position].relation]
                Log.e("RELATION",relations[nominees!![position].relation])
                nomineeText!!.setTextColor(resources.getColor(R.color.colorPrimaryDark))
                nomineeText!!.setPadding(0, 4, 0, 4)
                viewModel!!.nominee = nominees!![position]
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == Constants.PermissionAndRequests.PICK_IMAGE_REQUEST && resultCode == Activity.RESULT_OK && data != null) {
            val selectedImage = data.data
            val filePathColumn = arrayOf(MediaStore.Images.Media.DATA)

            val cursor = activity!!.contentResolver.query(selectedImage!!, filePathColumn, null, null, null)
            cursor!!.moveToFirst()

            val columnIndex = cursor.getColumnIndex(filePathColumn[0])
            val picturePath = cursor.getString(columnIndex)
            cursor.close()

            val bmp = BitmapFactory.decodeFile(picturePath)
            documentAddImage!!.setImageBitmap(bmp)
            documentAddImage!!.setPadding(0,0,0,0)
            documentAddText!!.visibility = View.GONE
            clickPolicy!!.visibility = View.GONE
            optional2!!.visibility = View.GONE
            optional1!!.text = "Policy Document"
            viewModel!!.bitmap = reducedImage(bmp)
        }
        if (requestCode == Constants.PermissionAndRequests.CAPTURE_IMAGE_REQUEST && resultCode == Activity.RESULT_OK && data != null) {
            val bmp = data.extras!!.get("data") as Bitmap
            documentAddImage!!.setImageBitmap(bmp)
            documentAddImage!!.setPadding(0,0,0,0)
            documentAddText!!.visibility = View.GONE
            clickPolicy!!.visibility = View.GONE
            optional2!!.visibility = View.GONE
            optional1!!.text = "Policy Document"
            viewModel!!.bitmap = reducedImage(bmp)
        }
    }

    private fun reducedImage(bmp: Bitmap): Bitmap {
        var bmp = bmp
        val stream = ByteArrayOutputStream()
        bmp.compress(Bitmap.CompressFormat.JPEG, 100, stream)
        val currSize = stream.toByteArray().size
        if (currSize > Constants.Documents.ONE_MEGABYTE) {
            val width = (bmp.width * Constants.Documents.ONE_MEGABYTE / currSize).toInt()
            val height = (bmp.height * Constants.Documents.ONE_MEGABYTE / currSize).toInt()
            bmp = Bitmap.createScaledBitmap(bmp, width, height, true)
        }
        return bmp
    }
}
